File: leetcode/problems/problem-specific/90.subsets-ii/90.subsets-ii.go
package main

import "sort"

func subsetsWithDup(nums []int) [][]int {
	if len(nums) <= 0 {
		return [][]int{}
	}

	res := [][]int{}
	sort.Ints(nums)
	subsetsWithDupHelper(nums, 0, []int{}, &res)
	return res
}

func subsetsWithDupHelper(nums []int, start int, subset []int, res *[][]int) {
	*res = append(*res, append([]int{}, subset...))

	for i := start; i < len(nums); i++ {
		if i > start && nums[i] == nums[i-1] {
			continue
		}

		subset = append(subset, nums[i])
		subsetsWithDupHelper(nums, i+1, subset, res)
		subset = subset[:len(subset)-1]
	}
}

File: leetcode/problems/problem-specific/90.subsets-ii/90.subsets-ii.py
#
# @lc app=leetcode id=90 lang=python3
#
# [90] Subsets II
#

# @lc code=start

# * Backtracking Solution | O(n*(2^n)) Time | O(n) Space
# * n -> The length of nums array

class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        result = []
        nums.sort()

        def subsetsWithDupHelper(start=0, subset=[]):
            result.append(subset[:])

            for i in range(start, len(nums)):
                if i > start and nums[i] == nums[i - 1]:
                    continue

                subset.append(nums[i])
                subsetsWithDupHelper(i + 1, subset)
                subset.pop()

        subsetsWithDupHelper()
        return result

# @lc code=end
